File: mobile/ios/Sources/RTCstackUI/Theme.swift
import SwiftUI

/// RTCstack design tokens, ported from `packages/ui-react/src/styles.css`.
/// Dark is the default (matches the web kit's `:root`); light mirrors `[data-theme="light"]`.
public struct RTCstackColors {
    public let bg: Color
    public let surface1: Color
    public let surface2: Color
    public let border: Color
    public let borderSubtle: Color
    public let text: Color
    public let textMuted: Color
    public let accent: Color
    public let danger: Color
    public let success: Color
    public let speakingRing: Color
    public let captionBg: Color
    public let captionText: Color

    public static let dark = RTCstackColors(
        bg: Color(hex: 0x0F0F0F),
        surface1: Color(hex: 0x1A1A1A),
        surface2: Color(hex: 0x252525),
        border: Color(hex: 0x333333),
        borderSubtle: Color(hex: 0x2A2A2A),
        text: Color(hex: 0xF5F5F5),
        textMuted: Color(hex: 0x999999),
        accent: Color(hex: 0x4F9CF9),
        danger: Color(hex: 0xE74C3C),
        success: Color(hex: 0x2ECC71),
        speakingRing: Color(hex: 0x4F9CF9),
        captionBg: Color.black.opacity(0.72),
        captionText: .white
    )

    public static let light = RTCstackColors(
        bg: Color(hex: 0xF5F5F5),
        surface1: .white,
        surface2: Color(hex: 0xEEEEEE),
        border: Color(hex: 0xDDDDDD),
        borderSubtle: Color(hex: 0xE5E5E5),
        text: Color(hex: 0x111111),
        textMuted: Color(hex: 0x666666),
        accent: Color(hex: 0x4F9CF9),
        danger: Color(hex: 0xE74C3C),
        success: Color(hex: 0x2ECC71),
        speakingRing: Color(hex: 0x4F9CF9),
        captionBg: Color.black.opacity(0.72),
        captionText: .white
    )
}

public enum RTCstackRadius {
    public static let sm: CGFloat = 4
    public static let md: CGFloat = 8
    public static let lg: CGFloat = 12
}

private struct RTCstackColorsKey: EnvironmentKey {
    static let defaultValue = RTCstackColors.dark
}

public extension EnvironmentValues {
    var rtcColors: RTCstackColors {
        get { self[RTCstackColorsKey.self] }
        set { self[RTCstackColorsKey.self] = newValue }
    }
}

public extension View {
    /// Provide RTCstack tokens to the view tree. Defaults to the current color scheme.
    func rtcstackTheme(_ colorScheme: ColorScheme? = nil) -> some View {
        modifier(RTCstackThemeModifier(forced: colorScheme))
    }
}

private struct RTCstackThemeModifier: ViewModifier {
    let forced: ColorScheme?
    @Environment(\.colorScheme) private var system

    func body(content: Content) -> some View {
        let scheme = forced ?? system
        let colors = scheme == .light ? RTCstackColors.light : RTCstackColors.dark
        content.environment(\.rtcColors, colors)
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255,
            opacity: 1
        )
    }
}

File: mobile/android/ui-compose/src/main/kotlin/com/rtcstack/ui/Theme.kt
package com.rtcstack.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

/**
 * RTCstack design tokens, ported from `packages/ui-react/src/styles.css`.
 * Dark is the default theme (matches the web kit's `:root`); light mirrors `[data-theme="light"]`.
 */
public data class RTCstackColors(
    val bg: Color,
    val surface1: Color,
    val surface2: Color,
    val border: Color,
    val borderSubtle: Color,
    val text: Color,
    val textMuted: Color,
    val accent: Color,
    val danger: Color,
    val success: Color,
    val speakingRing: Color,
    val captionBg: Color,
    val captionText: Color,
)

public object RTCstackTokens {
    public val dark: RTCstackColors = RTCstackColors(
        bg = Color(0xFF0F0F0F),
        surface1 = Color(0xFF1A1A1A),
        surface2 = Color(0xFF252525),
        border = Color(0xFF333333),
        borderSubtle = Color(0xFF2A2A2A),
        text = Color(0xFFF5F5F5),
        textMuted = Color(0xFF999999),
        accent = Color(0xFF4F9CF9),
        danger = Color(0xFFE74C3C),
        success = Color(0xFF2ECC71),
        speakingRing = Color(0xFF4F9CF9),
        captionBg = Color(0xB8000000),
        captionText = Color(0xFFFFFFFF),
    )

    public val light: RTCstackColors = dark.copy(
        bg = Color(0xFFF5F5F5),
        surface1 = Color(0xFFFFFFFF),
        surface2 = Color(0xFFEEEEEE),
        border = Color(0xFFDDDDDD),
        borderSubtle = Color(0xFFE5E5E5),
        text = Color(0xFF111111),
        textMuted = Color(0xFF666666),
    )

    public object Radius {
        public val sm: androidx.compose.ui.unit.Dp = 4.dp
        public val md: androidx.compose.ui.unit.Dp = 8.dp
        public val lg: androidx.compose.ui.unit.Dp = 12.dp
    }
}

public val LocalRTCstackColors: androidx.compose.runtime.ProvidableCompositionLocal<RTCstackColors> =
    staticCompositionLocalOf { RTCstackTokens.dark }

/** Convenience accessor: `RTCstackTheme.colors.accent`. */
public object RTCstackTheme {
    public val colors: RTCstackColors
        @Composable get() = LocalRTCstackColors.current
}

@Composable
public fun RTCstackTheme(
    darkTheme: Boolean = isSystemInDarkTheme(),
    content: @Composable () -> Unit,
) {
    val colors = if (darkTheme) RTCstackTokens.dark else RTCstackTokens.light
    val scheme = if (darkTheme) {
        darkColorScheme(
            primary = colors.accent,
            background = colors.bg,
            surface = colors.surface1,
            error = colors.danger,
            onBackground = colors.text,
            onSurface = colors.text,
        )
    } else {
        lightColorScheme(
            primary = colors.accent,
            background = colors.bg,
            surface = colors.surface1,
            error = colors.danger,
            onBackground = colors.text,
            onSurface = colors.text,
        )
    }
    CompositionLocalProvider(LocalRTCstackColors provides colors) {
        MaterialTheme(
            colorScheme = scheme,
            shapes = MaterialTheme.shapes.copy(
                small = RoundedCornerShape(RTCstackTokens.Radius.sm),
                medium = RoundedCornerShape(RTCstackTokens.Radius.md),
                large = RoundedCornerShape(RTCstackTokens.Radius.lg),
            ),
            content = content,
        )
    }
}
